src/categorical_algebra/CSets.jl

```diff
@@ -1429,6 +1433,271 @@ end
     end...))
 end
 
+
+"""A map f (from A to B) as a map of subobjects of A to subjects of B"""
+(f::ACSetTransformation)(X::SubACSet) = begin
+  codom(hom(X)) == dom(f) || error("Cannot apply $f to $X")
+  Subobject(codom(f); Dict(map(ob(acset_schema(dom(f)))) do o
+    o => sort(unique(f[o].(collect(components(X)[o]))))
+  end)...)
+end
+
+
+"""
+A map f (from A to B) as a map from A to a subobject of B
+# i.e. get the image of f as a subobject of B
+"""
+(f::ACSetTransformation)(X::StructACSet) =
+  X == dom(f) ? f(top(X)) : error("Cannot apply $f to $X")
+
+"""    preimage(f::ACSetTransformation,Y::Subobject)
+Inverse of f (from A to B) as a map of subobjects of B to subjects of A.
+"""
+function preimage(f::ACSetTransformation,Y::SubACSet)
+  codom(hom(Y)) == codom(f) || error("Cannot apply $f to $X")
+  Subobject(dom(f); Dict{Symbol, Vector{Int}}(map(ob(acset_schema(dom(f)))) do o
+    o => sort(unique(vcat([preimage(f[o],y) for y in collect(components(Y)[o])]...)))
+  end)...)
+end
+
+"""    preimage(f::CSetTransformation,Y::StructACSet)
+Inverse f (from A to B) as a map from subobjects of B to subobjects of A.
+Cast an ACSet to subobject, though this has a trivial answer when computing
+the preimage (it is necessarily the top subobject of A).
+"""
+preimage(f::CSetTransformation,Y::StructACSet) =
+  Y == codom(f) ? top(dom(f)) : error("Cannot apply inverse of $f to $Y")
+
+# VarACSets
+###########
+
+"""
+For any ACSet, X, a canonical map A→X where A has distinct variables for all
+subparts.
+"""
+function abstract(X::ACSet)
+  S = acset_schema(X)
+  A = deepcopy(X); 
+  comps = Dict{Any,Any}(map(attrtypes(S)) do at
+    rem_parts!(A, at, parts(A,at))
+    comp = Union{AttrVar,attrtype_type(X,at)}[]
+    for (f, d, _) in attrs(S; to=at)
+      append!(comp, A[f])
+      A[f] = AttrVar.(add_parts!(A, at, nparts(A,d)))
+    end 
+    at => comp
+  end)
+  for o in ob(S) comps[o]=parts(X,o) end
+  res = ACSetTransformation(A,X; comps...)
+  return res
+end 
+
+# Maximum Common C-Set
+######################
+
+""" Compute the size of a C-Set
+
+Defintion: let 𝐺: C → 𝐒et be a C-set, we define the _size_ of 𝐺 to be ∑_{c ∈ C}
+|𝐺c|.  For example, under this definition, the size of:
+  * a graph G is |GE| + |GV| (num edges + num vertices)
+  * a Petri net P is |PT| + |PS| + |PI| + |PO| (num transitions + num species +
+    num input arcs + num output arcs).
+"""
+size(X::ACSet) = mapreduce(oₛ -> nparts(X, oₛ), +, objects(acset_schema(X)); init=0)
+size(X::ACSetTransformation) = size(dom(X)) 
+
+"""
+Enumerate subobjects of an ACSet in order of largest to smallest 
+(assumes no wandering variables).
+"""
+struct SubobjectIterator
+  top::ACSet
+end
+
+function Base.collect(it::SubobjectIterator)
```

Comment:
Are you sure that this method needs to be implemented? I would expect `collect` to work automatically for any iterable, and the [iterators interface](https://docs.julialang.org/en/v1/manual/interfaces/) doesn't mention it.

Reply:
I looked into this and it appears that there is a much cleaner way of doing this. I have fixed our iterator definition to properly use the Iterator interface. Basically we just needed to tell the interface that the size is unknown. Apparently by default the iterators interface sets the size as "HasSize" but then if you try to get the size it says "size unknown". This seems like a pretty bad default for "optional" fields but that's a separate conversation outside of this repo.